uwblueprint/feeding-canadian-kids
# GraphQL query to retrieve all meal requests by requestor ID

## Notion Ticket Link

[Create Query for getting all food requests for one user](https://www.notion.so/uwblueprintexecs/Create-Query-for-getting-all-food-requests-for-one-user-e1effb42d80540bb8923ed9315e0bdf1?pvs=4)



## Implementation Description
* The query is used to retrieve all meal requests made by an ASP, using their user ID when executing the query
* Several query parameters can be used when executing the query: requestor ID (required), minimum drop off date, maximum drop off date, status ("Open", "Fulfilled", "Cancelled"), offset, limit, and sort by date direction ("ASC", "DESC"). These query parameters will be used when in the calendar view, for filtering and pagination.


## Steps To Test
1. From the root directory, run `make betest` and ensure that all automated tests pass (no regressions)

- [ ] All tests should pass (i.e. no regressions)

2. In the GraphQL playground, execute a mutation to add a new meal request. You can use the one below:

```
mutation {
  createMealRequest(description: "test description", requestorId: "646fd88671a4b59e02a20d8d", dropOffTime: "02:03:39", dropOffLocation:"location", requestDates:["2023-10-23"], mealInfo:{portions:2}, onsiteStaff:{name:"John Doe", email:"test@gmail.com", phone:"1234567890"}) {
    mealRequests {
      id,
    }
  }
}
```

- [ ] The mutation should succeed, adding a new document to the `meal_requests` collection, and there should be no errors thrown.

3. In the GraphQL playground, run a query for all meal requests from a specific `requestorId` (you can use one of the following IDs: `507f1f77bcf86cd799439029`, `64a49d1419ff1c5bb3bb84b3`, `646fd88671a4b59e02a20d8d`. e.g.

```
{
  getMealRequestsByRequestorId(requestorId: "507f1f77bcf86cd799439029") {
    id,
    requestor {
      id
    },
    description,
    status,
    dropOffDatetime,
    dropOffLocation,
    mealInfo {
      portions
      dietaryRestrictions
      mealSuggestions
    },
    onsiteStaff {
    	name
      email
      phone
    },
    dateCreated,
    dateUpdated,
    deliveryInstructions,
    donationInfo {
      donor {
        id
      },
      commitmentDate
      mealDescription
      additionalInfo
    }
  }
}
```

- [ ] All meal requests returned should be from the `requestor` that you specified.

4. Run a query specifying values for `minDropOffDate` and `maxDropOffDate`. e.g.

```
{
  getMealRequestsByRequestorId(requestorId: "507f1f77bcf86cd799439029", minDropOffDate: "2023-09-01", maxDropOffDate: "2023-09-15") {
    id,
    requestor {
      id
    },
    dropOffDatetime
  }
}
```

- [ ] The `dropOffDatetime` for all returned meal requests should fall within the range specified.

5. Run a query specifying a value or values for the `status` list (`OPEN`, `FULFILLED`, `CANCELLED`). e.g.

```
{
  getMealRequestsByRequestorId(requestorId: "507f1f77bcf86cd799439029", status: [OPEN]) {
    id,
    requestor {
      id
    },
    status
  }
}
```

- [ ] The `status` for all returned meal requests should be in the list of statuses specified.

6. Run a query specifying the `sortByDateDirection` (`ASCENDING` or `DESCENDING`). e.g.

```
{
  getMealRequestsByRequestorId(requestorId: "507f1f77bcf86cd799439029", sortByDateDirection: DESCENDING) {
    id,
    requestor {
      id
    },
    dateCreated
  }
}
```

- [ ] The returned meal requests should be ordered by `dateCreated` in either ascending or descending order, whichever was specified.

7. Run a query specifying the `offset` and `limit`. e.g.

```
{
  getMealRequestsByRequestorId(requestorId: "507f1f77bcf86cd799439029", offset: 2, limit: 2) {
    id,
    requestor {
      id
    }
  }
}
```

- [ ] The number of returned meal requests should match the parameters that were specified.


## What Should Reviewers Focus On?
* Comments were left in the PR for areas of interest.


## Checklist
- [-] I have added tests for my changes - will be added in a follow-up PR
